### ISSUE TITLE
AlphaFold 3

### ISSUE THREAD

**Issue by laraPPr** [vscentrum/vsc-software-stack#477]
* link to support ticket: [2024120960000091](https://otrsdict.ugent.be/otrs/index.pl?Action=AgentTicketZoom;TicketID=178669) + [2025091160000376](https://otrsdict.ugent.be/otrs/index.pl?Action=AgentTicketZoom;TicketID=210750) + [2025100760000354](https://otrsdict.ugent.be/otrs/index.pl?Action=AgentTicketZoom;TicketID=214711)
* website: https://github.com/google-deepmind/alphafold3 
* installation docs: https://github.com/google-deepmind/alphafold3/blob/main/docs/installation.md
* toolchain: `foss/2024a?`
* easyblock to use: `PythonBundle`
* required dependencies:
  * [ ] https://github.com/google-deepmind/alphafold3/blob/main/requirements.txt
* optional dependencies:
  * [ ] ...
* notes:
  * ...
* effort: *(TBD)*
* other install methods
  * conda: yes (link?) / no
  * container image: yes (link?) / no
  * pre-built binaries (RHEL8 Linux x86_64): yes (link?) / no
  * easyconfig outside EasyBuild: yes / no

**Comment by lexming** [comment]
Duplicate of #454?

**Comment by laraPPr** [comment]
We can collapse them if we need to. But would also like to set our inuits people on it as well.

**Comment by laraPPr** [comment]
Someone already started on an EasyConfig https://github.com/easybuilders/easybuild-easyconfigs/pull/21864

**Comment by boegel** [comment]
@pavelToman Can you look into testing https://github.com/easybuilders/easybuild-easyconfigs/pull/22028 (submit test report), and figure out what needs to happen to get https://github.com/easybuilders/easybuild-easyconfigs/pull/21924 across the finish line?

And then maybe also talk to @ThomasHoffmann77 and figure out what else needs to happen to get AlphaFold 3 fully supported in EasyBuild...

**Comment by ThomasHoffmann77** [comment]
> @pavelToman Can you look into testing [easybuilders/easybuild-easyconfigs#22028](https://github.com/easybuilders/easybuild-easyconfigs/pull/22028) (submit test report), and figure out what needs to happen to get [easybuilders/easybuild-easyconfigs#21924](https://github.com/easybuilders/easybuild-easyconfigs/pull/21924) across the finish line?
> 
> And then maybe also talk to @ThomasHoffmann77 and figure out what else needs to happen to get AlphaFold 3 fully supported in EasyBuild...

- I moved the checksums in  https://github.com/easybuilders/easybuild-easyconfigs/pull/22028 tensorstore
- current progress in AF3 easyconfig: see also https://github.com/easybuilders/easybuild-easyconfigs/pull/22067
- I am currently struggling with Clang sanity checks for Triton 3.1.0 (https://github.com/easybuilders/easybuild-easyconfigs/pull/22064). For test purpose I omitted the sanity check step for Clang 19.0.0 at our site in order to get Triton built. 

@pavelToman @boegel Update: I switched to the latest AlphaFold3 commit to fix some problems. I got the module built at our site. However, at a testrun* it turned out, that https://github.com/easybuilders/easybuild-easyconfigs/pull/21924 is missing jax_cuda12_plugin and jax_cuda12_pjrt. I am working on it.
Furthermore, the jax module is still missing the test-step and bazel downloads need to be switched off.

*) AF3 testrun of the data pipeline ran through. Inference step failed.

**Comment by boegel** [comment]
@pavelToman Can you look into the easyconfig PRs that @ThomasHoffmann77 mentioned? (quick review + test)

**Comment by ThomasHoffmann77** [comment]
> [@pavelToman](https://github.com/pavelToman) Can you look into the easyconfig PRs that [@ThomasHoffmann77](https://github.com/ThomasHoffmann77) mentioned? (quick review + test)

@boegel @pavelToman regarding jax: the problem with the missing jax_cuda_{plugin,pjrt} has been fixed. I also ran all tests of which 5 were failing.
Recently I was working on an update of this PR for offline build pr-downloading the complete set of bazel artifact into a dist directory. I am currently struggling with replacing hermetic python by EB python. Furthermore, I did not find a way to avoid some downloads of python packages, yet (-> patch rules_python).
I did not upload my latest work, yet.

The jax(lib) easyblock might require an update to get things less complicated. Maybe @Flamefire could help here?

Maybe it would also make sense to switch to the latest jax (0.5.0) (?)

**Comment by Flamefire** [comment]
I don't have much experience with jax and haven't built TensorFlow for quite a while now. But I can see what I can do based on what you got already. I think we have quite a bit of experience bottled in our existing ECs using Bazel.   
I'm a bit busy right now so might not be able to do much soon

**Comment by laraPPr** [comment]
We have a new request come in from anothe user. I'll add it to the description.

**Comment by pavelToman** [comment]
PR:
- https://github.com/easybuilders/easybuild-easyconfigs/pull/24177

**Comment by laraPPr** [comment]
#### Sumarry
still waiting on:
- https://github.com/easybuilders/easybuild-easyconfigs/pull/24177

**Comment by laraPPr** [comment]
@pavelToman  Can you make an estimate of how much time (# hours) this request took you?

**Comment by pavelToman** [comment]
took 37h

**Comment by laraPPr** [comment]
PR merged, software installed, ready for cleanup

**Comment by pavelToman** [comment]
TODO:
- update to foss-2025a/b
- update to v3.0.4
- https://github.com/vscentrum/vsc-software-stack/issues/454#issuecomment-5521241611
- Kenneth's Draft PR: https://github.com/easybuilders/easybuild-easyconfigs/pull/26779

**Comment by pavelToman** [comment]
The update to the foss/2025b and v3.0.4 requires to create/update ECs for deps:
- jax-0.10.2-CUDA-12.9.1 
- jaxtyping-0.3.5-CUDA-12.9.1
- dm-haiku-0.0.16-CUDA-12.9.1
- dssp-4.4.10
- zstandard-0.25.0
- typeguard-2.13.3
- etils-1.13.0
- tokamax-0.0.12
- tensorboardX-2.6.5
- libcifpp, mrc, libmcfp,

**Comment by pavelToman** [comment]
PRs deps:
- https://github.com/easybuilders/easybuild-easyconfigs/pull/26843
- https://github.com/easybuilders/easybuild-easyconfigs/pull/26844

**Comment by pavelToman** [comment]
there is big issue for AlphaFold 3.0.4 and toolchain foss/2025b:
- AF3 requires jax==0.10.2
- jax 0.10.2 needs NCCL 2.29.7
- for foss/2025b there is NCCL 2.27.7 which is too old to satisfy jax

**Comment by boegel** [comment]
> there is big issue for AlphaFold 3.0.4 and toolchain foss/2025b:
> 
>     * AF3 requires jax==0.10.2
> 
>     * jax 0.10.2 needs NCCL 2.29.7
> 
>     * for foss/2025b there is NCCL 2.27.7 which is too old to satisfy jax

@pavelToman Can you check if `jax` 0.10.2 *really* needs NCCL 2.29, i.e. if it requires a particular feature that is not yet available in NCCL 2.27.7?
Did you try building `jax` 0.10.2 on top of NCCL 2.27? If so, what problems did you run into?

Same question for AlphaFold: does it *really* need jax>=0.10.2, or can we just lower that requirement to the jax 0.8.1 we have for `2025b`?

**Comment by pavelToman** [comment]
> > there is big issue for AlphaFold 3.0.4 and toolchain foss/2025b:
> > ```
> > * AF3 requires jax==0.10.2
> > 
> > * jax 0.10.2 needs NCCL 2.29.7
> > 
> > * for foss/2025b there is NCCL 2.27.7 which is too old to satisfy jax
> > ```
> 
> [@pavelToman](https://github.com/pavelToman) Can you check if `jax` 0.10.2 _really_ needs NCCL 2.29, i.e. if it requires a particular feature that is not yet available in NCCL 2.27.7? Did you try building `jax` 0.10.2 on top of NCCL 2.27? If so, what problems did you run into?
> 
> Same question for AlphaFold: does it _really_ need jax>=0.10.2, or can we just lower that requirement to the jax 0.8.1 we have for `2025b`?

Yes, I tried building JAX 0.10.2 with NCCL 2.27.7.
The XLA used by JAX 0.10.2 does require functionality that is not available in NCCL 2.27.7.
The build with NCCL 2.27.7 fails while compiling `xla/stream_executor/cuda/ragged_all_to_all_kernel_cuda.cc` with:
`fatal error: 'third_party/nccl/nccl_device.h' file not found`
This API was introduced in NCCL 2.28, so 2.27.7 is definitely too old for this XLA.
JAX 0.10.2's own CUDA Bazel configuration defaults to NCCL 2.29.7, which is why I am considering 2.29.7.

AlphaFold3 v3.0.4 does not specify jax>=0.10.2; it pins it exactly:
jax==0.10.2 and jax[cuda12]==0.10.2
This is probably because of XLA - AF3 v3.0.4 needs the Tokamax and this depends closely on JAX/Pallas/XLA internals.

Also AF3 v3.0.3 has pinned dependency on jax==0.9.1, so our jax-0.8.1 is too old either for AF3 v3.0.3.

I will create new NCCL and jax and these will be only deps for AF3. We can have either jax-0.8.1-CUDA which can be a dependency for other software and also compatible with PyTorch, since this version should be ok with NCCL 2.27.7.

Also it seems I can build jax-0.10.2 without the NCCL because the AF3 does not use NCCL.
